File: multiitemkit/src/main/java/com/sanniou/multiitemkit/decoration/HorizontalItemDecoration.java
package com.sanniou.multiitemkit.decoration;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import android.view.View;

import androidx.annotation.DrawableRes;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class HorizontalItemDecoration extends RecyclerView.ItemDecoration {

  private static final int[] ATTRS = {android.R.attr.listDivider};

  private final Map<Integer, Drawable> mDividerViewTypeMap;
  private final Drawable mFirstDrawable;
  private final Drawable mLastDrawable;
  private final Drawable mCommonDrawable;
  private final boolean mDrawOnTop;

  public HorizontalItemDecoration(Map<Integer, Drawable> dividerViewTypeMap,
                                  Drawable firstDrawable, Drawable lastDrawable, Drawable commonDrawable, boolean drawOnTop) {
    mDividerViewTypeMap = dividerViewTypeMap;
    mFirstDrawable = firstDrawable;
    mLastDrawable = lastDrawable;
    mCommonDrawable = commonDrawable;
    mDrawOnTop = drawOnTop;
  }

  @Override
  public void getItemOffsets(Rect outRect, View view, RecyclerView parent,
                             RecyclerView.State state) {
    // specific view type
    int childType = parent.getLayoutManager().getItemViewType(view);

    // last position
    if (isLastPosition(view, parent)) {
      if (mLastDrawable != null) {
        outRect.right = mLastDrawable.getIntrinsicWidth();
      }
      if (mDrawOnTop) {
        Drawable drawableEnd = mDividerViewTypeMap.get(childType);
        if (drawableEnd == null) {
          drawableEnd = mCommonDrawable;
        }
        outRect.left = drawableEnd.getIntrinsicWidth();
      }
      return;
    }

    Drawable drawable = mDividerViewTypeMap.get(childType);
    if (drawable == null) {
      drawable = mCommonDrawable;
    }
    if (drawable != null) {
      if (mDrawOnTop) {
        outRect.left = drawable.getIntrinsicWidth();
      } else {
        outRect.right = drawable.getIntrinsicWidth();
      }
    }

    // first position
    if (isFirstPosition(view, parent)) {
      if (mFirstDrawable != null) {
        outRect.left = mFirstDrawable.getIntrinsicWidth();
      } else if (mDrawOnTop) {
        Drawable drawableStart = mDividerViewTypeMap.get(childType);
        if (drawableStart == null) {
          drawableStart = mCommonDrawable;
        }
        outRect.left = drawableStart.getIntrinsicWidth();
      }
    }
  }

  @Override
  public void onDraw(Canvas c, RecyclerView parent, RecyclerView.State state) {
    int top = parent.getPaddingTop();
    int bottom = parent.getHeight() - parent.getPaddingBottom();

    int childCount = parent.getChildCount();
    for (int i = 0; i <= childCount - 1; i++) {
      View child = parent.getChildAt(i);
      int childViewType = parent.getLayoutManager().getItemViewType(child);
      RecyclerView.LayoutParams params = (RecyclerView.LayoutParams) child.getLayoutParams();
      // last position
      if (isLastPosition(child, parent)) {
        if (mDrawOnTop) {
          Drawable drawableEnd = mDividerViewTypeMap.get(childViewType);
          if (drawableEnd == null) {
            drawableEnd = mCommonDrawable;
          }
          if (drawableEnd != null) {
            int left, right;
            right = child.getLeft() - params.leftMargin;
            left = right - drawableEnd.getIntrinsicWidth();
            drawableEnd.setBounds(left, top, right, bottom);
            drawableEnd.draw(c);
          }
        }
        if (mLastDrawable != null) {
          int left, right;
          left = child.getRight() + params.rightMargin;
          right = left + mLastDrawable.getIntrinsicWidth();
          mLastDrawable.setBounds(left, top, right, bottom);
          mLastDrawable.draw(c);
        }
        return;
      }

      // specific view type
      Drawable drawable = mDividerViewTypeMap.get(childViewType);
      if (drawable == null) {
        drawable = mCommonDrawable;
      }
      if (drawable != null) {
        int left, right;
        if (mDrawOnTop) {
          right = child.getLeft() - params.leftMargin;
          left = right - drawable.getIntrinsicWidth();
        } else {
          left = child.getRight() + params.rightMargin;
          right = left + drawable.getIntrinsicWidth();
        }
        drawable.setBounds(left, top, right, bottom);
        drawable.draw(c);
      }

      // first position
      if (isFirstPosition(child, parent) && mFirstDrawable != null) {

        int left, right;
        if (mDrawOnTop) {
          Drawable drawableStart = mDividerViewTypeMap.get(childViewType);
          if (drawableStart == null) {
            drawableStart = mCommonDrawable;
          }
          left = child.getRight() + params.rightMargin;
          right = left + drawableStart.getIntrinsicWidth();
        } else {
          right = child.getLeft() - params.leftMargin;
          left = right - mFirstDrawable.getIntrinsicWidth();
        }
        mFirstDrawable.setBounds(left, top, right, bottom);
        mFirstDrawable.draw(c);
      }
    }
  }

  private boolean isFirstPosition(View view, RecyclerView parent) {
    return parent.getChildAdapterPosition(view) == 0;
  }

  private boolean isLastPosition(View view, RecyclerView parent) {
    return parent.getChildAdapterPosition(view) == parent.getAdapter().getItemCount() - 1;
  }

  public static class Builder {

    private final Context mContext;
    private final Map<Integer, Drawable> mDividerViewTypeMap = new HashMap<>();
    private Drawable mFirstDrawable;
    private Drawable mLastDrawable;
    private Drawable mCommonDrawable;
    private boolean mDrawOnTop;

    public Builder(Context context) {
      mContext = context;
    }

    public Builder drawOnTop(boolean drawOnTop) {
      mDrawOnTop = drawOnTop;
      return this;
    }

    public Builder common(Drawable drawable) {
      mCommonDrawable = drawable;
      return this;
    }

    public Builder type(List<Integer> types, Drawable drawable) {
      if (types == null || drawable == null) {
        return this;
      }
      for (Integer type : types) {
        type(type, drawable);
      }
      return this;
    }

    public Builder type(int viewType) {
      final TypedArray a = mContext.obtainStyledAttributes(ATTRS);
      Drawable divider = a.getDrawable(0);
      type(viewType, divider);
      a.recycle();
      return this;
    }

    public Builder type(int viewType, @DrawableRes int drawableResId) {
      mDividerViewTypeMap.put(viewType, ContextCompat.getDrawable(mContext, drawableResId));
      return this;
    }

    public Builder type(int viewType, Drawable drawable) {
      mDividerViewTypeMap.put(viewType, drawable);
      return this;
    }

    public Builder first(@DrawableRes int drawableResId) {
      first(ContextCompat.getDrawable(mContext, drawableResId));
      return this;
    }

    public Builder first(Drawable drawable) {
      mFirstDrawable = drawable;
      return this;
    }

    public Builder last(@DrawableRes int drawableResId) {
      last(ContextCompat.getDrawable(mContext, drawableResId));
      return this;
    }

    public Builder last(Drawable drawable) {
      mLastDrawable = drawable;
      return this;
    }

    public HorizontalItemDecoration create() {
      return new HorizontalItemDecoration(mDividerViewTypeMap, mFirstDrawable,
          mLastDrawable, mCommonDrawable,
          mDrawOnTop);
    }
  }
}

File: multiitemkit/src/main/java/com/sanniou/multiitemkit/decoration/HorizontalItemDecoration.kt
package com.sanniou.multiitemkit.decoration

import android.content.Context
import android.graphics.Canvas
import android.graphics.Rect
import android.graphics.drawable.Drawable
import android.util.SparseArray
import android.view.View
import androidx.annotation.DrawableRes
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ItemDecoration

class HorizontalItemDecoration(
    private val mDividerViewTypeMap: SparseArray<Drawable?>,
    private val mFirstDrawable: Drawable?,
    private val mLastDrawable: Drawable?,
    private val mCommonDrawable: Drawable?,
    private val mDrawOnTop: Boolean
) : ItemDecoration() {
    override fun getItemOffsets(
        outRect: Rect, view: View, parent: RecyclerView,
        state: RecyclerView.State
    ) { // specific view type
        val childType = parent.layoutManager?.getItemViewType(view) ?: return

        // first position
        if (isFirstPosition(view, parent)) {
            if (mFirstDrawable != null) {
                outRect.left = mFirstDrawable.intrinsicWidth
            } else if (mDrawOnTop) {
                val drawableStart = mDividerViewTypeMap[childType] ?: mCommonDrawable ?: return
                outRect.left = drawableStart.intrinsicWidth
            }
            return
        }

        // last position
        if (isLastPosition(view, parent)) {
            if (mLastDrawable != null) {
                outRect.right = mLastDrawable.intrinsicWidth
            }
            if (mDrawOnTop) {
                val drawableEnd = mDividerViewTypeMap[childType] ?: mCommonDrawable ?: return
                outRect.left = drawableEnd.intrinsicWidth
            }
            return
        }
        val drawable = mDividerViewTypeMap[childType] ?: mCommonDrawable ?: return

        if (mDrawOnTop) {
            outRect.left = drawable.intrinsicWidth
        } else {
            outRect.right = drawable.intrinsicWidth
        }
    }

    override fun onDraw(
        c: Canvas,
        parent: RecyclerView,
        state: RecyclerView.State
    ) {
        val top = parent.paddingTop
        val bottom = parent.height - parent.paddingBottom
        val childCount = parent.childCount
        var left = 0
        var right = 0
        for (i in 0 until childCount) {
            val child = parent.getChildAt(i)
            val childViewType = parent.layoutManager?.getItemViewType(child) ?: return
            val params =
                child.layoutParams as RecyclerView.LayoutParams

            // first position
            if (isFirstPosition(child, parent)) {
                if (mFirstDrawable != null) {
                    right = child.left - params.leftMargin
                    left = right - mFirstDrawable.intrinsicWidth
                    mFirstDrawable.setBounds(left, top, right, bottom)
                    mFirstDrawable.draw(c)
                } else if (mDrawOnTop) {
                    val drawableStart =
                        mDividerViewTypeMap[childViewType] ?: mCommonDrawable ?: return
                    left = child.right + params.rightMargin
                    right = left + drawableStart.intrinsicWidth
                    drawableStart.setBounds(left, top, right, bottom)
                    drawableStart.draw(c)
                }
                return
            }

            // last position
            if (isLastPosition(child, parent)) {
                if (mLastDrawable != null) {
                    left = child.right + params.rightMargin
                    right = left + mLastDrawable.intrinsicWidth
                    mLastDrawable.setBounds(left, top, right, bottom)
                    mLastDrawable.draw(c)
                } else if (mDrawOnTop) {
                    val drawableEnd =
                        mDividerViewTypeMap[childViewType] ?: mCommonDrawable ?: return
                    right = child.left - params.leftMargin
                    left = right - drawableEnd.intrinsicWidth
                    drawableEnd.setBounds(left, top, right, bottom)
                    drawableEnd.draw(c)
                }

                return
            }
            // specific view type
            val drawable = mDividerViewTypeMap[childViewType] ?: mCommonDrawable ?: return
            if (mDrawOnTop) {
                right = child.left - params.leftMargin
                left = right - drawable.intrinsicWidth
            } else {
                left = child.right + params.rightMargin
                right = left + drawable.intrinsicWidth
            }
            drawable.setBounds(left, top, right, bottom)
            drawable.draw(c)
        }
    }

    private fun isFirstPosition(view: View, parent: RecyclerView): Boolean {
        return parent.getChildAdapterPosition(view) == 0
    }

    private fun isLastPosition(view: View, parent: RecyclerView): Boolean {
        return parent.getChildAdapterPosition(view) == parent.adapter?.itemCount ?: 0 - 1
    }

    class Builder(private val mContext: Context) {
        private val mDividerViewTypeMap = SparseArray<Drawable?>()
        private var mFirstDrawable: Drawable? = null
        private var mLastDrawable: Drawable? = null
        private var mCommonDrawable: Drawable? = null
        private var mDrawOnTop = false
        fun drawOnTop(drawOnTop: Boolean): Builder {
            mDrawOnTop = drawOnTop
            return this
        }

        fun common(drawable: Drawable?): Builder {
            mCommonDrawable = drawable
            return this
        }

        fun type(
            types: List<Int>?,
            drawable: Drawable?
        ): Builder {
            if (types == null || drawable == null) {
                return this
            }
            for (type in types) {
                type(type, drawable)
            }
            return this
        }

        fun type(viewType: Int): Builder {
            val a =
                mContext.obtainStyledAttributes(ATTRS)
            val divider = a.getDrawable(0)
            type(viewType, divider)
            a.recycle()
            return this
        }

        fun type(viewType: Int, @DrawableRes drawableResId: Int): Builder {
            mDividerViewTypeMap.put(viewType, ContextCompat.getDrawable(mContext, drawableResId))
            return this
        }

        fun type(
            viewType: Int,
            drawable: Drawable?
        ): Builder {
            mDividerViewTypeMap.put(viewType, drawable)
            return this
        }

        fun first(@DrawableRes drawableResId: Int): Builder {
            first(ContextCompat.getDrawable(mContext, drawableResId))
            return this
        }

        fun first(drawable: Drawable?): Builder {
            mFirstDrawable = drawable
            return this
        }

        fun last(@DrawableRes drawableResId: Int): Builder {
            last(ContextCompat.getDrawable(mContext, drawableResId))
            return this
        }

        fun last(drawable: Drawable?): Builder {
            mLastDrawable = drawable
            return this
        }

        fun create(): HorizontalItemDecoration {
            return HorizontalItemDecoration(
                mDividerViewTypeMap, mFirstDrawable,
                mLastDrawable, mCommonDrawable,
                mDrawOnTop
            )
        }
    }

    companion object {
        private val ATTRS = intArrayOf(android.R.attr.listDivider)
    }
}